tests/components/roborock/test_config_flow.py

```diff
@@ -0,0 +1,88 @@
+"""Test Roborock config flow."""
+from unittest.mock import patch
+
+import pytest
+from roborock.exceptions import RoborockException
+
+from homeassistant import config_entries, data_entry_flow
+from homeassistant.components.roborock.const import CONF_ENTRY_CODE, DOMAIN
+from homeassistant.core import HomeAssistant
+
+from .mock_data import MOCK_CONFIG, USER_EMAIL
+
+
+@pytest.mark.parametrize(
+    (
+        "request_code_side_effect",
+        "request_code_errors",
+        "code_login_side_effect",
+        "code_login_errors",
+    ),
+    [
+        (None, {}, RoborockException(), {"base": "invalid_code"}),
+        (None, {}, Exception(), {"base": "unknown"}),
+        (RoborockException(), {"base": "invalid_email"}, None, {}),
+        (Exception(), {"base": "unknown"}, None, {}),
+    ],
+)
+async def test_config_flow_failures(
+    hass: HomeAssistant,
+    bypass_api_fixture,
+    request_code_side_effect: Exception | None,
+    request_code_errors: dict[str, str],
+    code_login_side_effect: Exception | None,
+    code_login_errors: dict[str, str],
+) -> None:
+    """Handle applying errors to request code or code login and recovering from the errors."""
+    with patch(
+        "homeassistant.components.roborock.async_setup_entry", return_value=True
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": config_entries.SOURCE_USER}
+        )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+    if request_code_side_effect:
```

Comment:
You've made good improvements here based on the fixtures, etc However one thing to further clarify is that we don't want conditionals in tests like this.  

So to be more specific:
- have one test that fails based on request code failures
- have one test that fails based on login code failures

The motivation behind this is we'd rather have some code duplication, but easier to follow tests.

I also think you want to add back a test that completes the whole flow without errors from the start.